Auto-create sample board menu option #1211

codemanmeet opened the pull request:
Fixes #1211. I have created a [sample repo](https://github.com/codemanmeet/HubTurboTest/issues) in my personal account as of now. The final sample repo in the HubTurbo organization would look like that.

damithc commented:
Can you list the panel names and filters here? We need to make sure they showcase the power of HT. Try to use filters not available in normal GitHub.

codemanmeet commented:
`Open issues and PR's` -> `repo:HubTurbo/SampleRepo (is:issue OR is:pr) is:open`
`V5 Milestone` -> `repo:HubTurbo/SampleRepo milestone:V5 sort:status`
`Urgent issues assigned to Darius` -> `repo:HubTurbo/SampleRepo label:\"urgent\" assignee:dariusf`

Currently it is the above!

@damithc Do you have any particular preferences of qualifiers that we could use?!

damithc commented:
We need to make sure they showcase the power of HT. **Try to use filters not available in normal GitHub**. Give it a try yourself first. Come up with a list of possible filters and the rest of us can give our views on which ones are better choices.

codemanmeet commented:
1. `Issues with labels but no milestones` -> `repo:HubTurbo/SampleRepo has:label NOT(has:milestones)`. This showcases the `NOT` as well has the `has` power of HubTurbo currently not available on Github

2. `Open issues and PR's` -> `repo:HubTurbo/SampleRepo (is: issue OR is:pr)`. `OR` is currently not available on Github.

3.`Issues assigned to Darius not closed in V4 or V5` -> `repo:HubTurbo/SampleRepo assignee:dariusf -state:closed (milestone:V5 OR milestone:V4)`. This showcases the NOT sign (`-`).

Let me know if this looks good! :)

damithc commented:
Go through our filter documentation and figure out all the unique things we have. I think you are missing many. Need to be more thorough.

codemanmeet commented:
Comparing Github's (filter documentation)[https://help.github.com/articles/searching-issues/], I found the following filter-related advantages of HubTurbo- 

1. Presence of `id` search in HubTurbo
2. Presence of `has` in HubTurbo
3. Usage of `AND`, `NOT` and `OR`
4. Multiple repository search
5. Advanced sorting, use of multiple sort item like `nonSelfUpdate`, `comments`, `labels` all combined in one qualifier and applied in that order
6. The use of negation (`-` / `~`) for all qualifiers, E.g. `-state:closed`
7. Use of aliases for qualifiers.

The rest of the qualifiers like `assignee`, `updated` already exist on Github.

The three panels in the Sample Board would have filters that cover all the 7 advantages of HubTurbo listed above - 

1. `Issues with labels but no milestones and ID>5` -> `repo:HubTurbo/SampleRepo has:label NOT(has:milestones) id:>5`.

2. `Open issues in my two sample repos` -> `(repo:HubTurbo/SampleRepo OR repo:HubTurbo/HubTurbo) is: issue is:open sort:priority,comments`.

3. `Issues assigned to Darius not closed in V4 or V5` -> `repo:HubTurbo/SampleRepo as:dariusf -state:closed (milestone:V5 OR milestone:V4)`.

dariusf commented:
Off the top of my head, a few more:

- We overload some qualifiers. `updated` can take a number of hours or a number range, for example.
- There are some dynamic qualifiers, like `milestone:current`. Might want to mention those under aliases.
- Shorthand syntax (`;` for OR is one example)
- Parentheses for grouping
- Quoting, allowing search terms with spaces

Also we should probably be consistent with the logical operators, like using either `-` or `NOT`.

codemanmeet commented:
I have collated a list of HubTurbo-specific filter related advantages. These are - 

1. Presence of `id` search in HubTurbo
2. Presence of `has` in HubTurbo
3. Usage of `AND`, `NOT` and `OR`
4. Multiple repository search
5. Advanced sorting, use of multiple sort item like `nonSelfUpdate`, `comments`, `labels` all combined in one qualifier and applied in that order
6. The use of negation (`-` / `~`) for all qualifiers, E.g. `-state:closed`
7. Use of aliases for qualifiers.
8. We overload some qualifiers. updated can take a number of hours or a number range, for example.
9. There are some dynamic qualifiers, like milestone:current. Might want to mention those under aliases.
10. Shorthand syntax (; for OR is one example)
11. Parentheses for grouping
12. Quoting, allowing search terms with spaces

@damithc As discussed, I'm listing down 10 possible filters to discuss the core advantages of HubTurbo. Kindly let me know which 3 you prefer since we can have only 3 panels in our sample board!

1. `Issues with labels but no milestones and ID>5` -> `repo:HubTurbo/SampleRepo has:label NOT(has:milestones) id:>5`.

2. `Open issues in my two sample repos` -> `(repo:HubTurbo/SampleRepo OR repo:HubTurbo/HubTurbo) is: issue is:open sort:priority,comments`. @dariusf This would require the creation of two sample repos. If not possible, we can use `HubTurbo/HubTurbo` as the second repo to demonstrate multiple repository search!

3. `Issues assigned to Darius not closed in V4 or V5` -> `repo:HubTurbo/SampleRepo as:dariusf NOT(state:closed) (milestone:V5 OR milestone:V4)`. 

4. `Open Issues in the current milestone` -> `repo:HubTurbo/SampleRepo milestone:current is:issue is:open`

5. `Issues updated between 2 Feb to 3 Feb 2016` -> `repo:HubTurbo/SampleRepo updated:2016-02-02 .. 2016-02-03 is:issue`

6. `Open issues in the that involve Darius or Damith` ->  `repo:HubTurbo/SampleRepo is:issue is:open (user:dariusf || user:damithc)`

7. `Unmerged PR's` -> `repo:HubTurbo/SampleRepo is:unmerged is:pr`

8. `Issues that contain the term "UI"` -> `repo:HubTurbo/SampleRepo sort:updated "UI"`

9. `Open Issues created on/after 1 Feb 2016` -> `repo:HubTurbo/SampleRepo is:issue is:open created:>=2016-2-1`

10. `Unread issues that are urgent` -> `repo:HubTurbo/SampleRepo is:issue is:open label:"urgent"`

codemanmeet commented:
@dariusf Let me know which filters you prefer as well :)

dariusf commented:
I think we should demonstrate real use cases, incidentally slipping the new features in. 10 (for `count`) and 2 (OR, parentheses, complex sorting) look good, and maybe 7 (multiple repos). We can use the shorthand syntax where possible.

damithc commented:
Agree with @dariusf about using realistic filters. 
@codemanmeet how come `is:read` not mentioned under unique features?
There is no need to limit to one sample repo. Being able to view issues from multiple repos is one of our strongest value propositions.

codemanmeet commented:
`is:read` is under point 8 in the feature list. I have expanded the description for that feature. I like 10, 2 and 7 too. I find 4 quite important and realistic too. Let's fix on 2, 7 and 10. Yup, I agree showcasing multiple repository search is quite important! There's a second sample repo created as well. I have made the relevant code changes. 
Issues in the first sample repo will look like - https://github.com/codemanmeet/HubTurboTest/issues. 
Issues in the second sample repo will look like - https://github.com/codemanmeet/HubTurboTest2/issues
Let me know if this can be merged :)

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -23,6 +23,20 @@
+            "repo:" + SECOND_SAMPLE_REPO_NAME + ")" + " " + "is: issue sort:comments,updated";
+    public static final String SECOND_SAMPLE_FILTER = "repo:" + FIRST_SAMPLE_REPO_NAME + " " +
+            "count:3 is:issue is:open label:\"urgent\"";
+    public static final String THIRD_SAMPLE_FILTER = "repo:" + FIRST_SAMPLE_REPO_NAME
```
Could we use an array to store these? Then we don't have to copy-paste for all the rest of the code. Also, format strings might be neater.

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -101,6 +119,28 @@ private void createWorkAllocationBoard() {
+
+        DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD,
+                SAMPLE_BOARD + " has been created and loaded.\n\n" +
+                        "It is saved under the name \"" + SAMPLE_BOARD + "\".");
```
Same here, use format strings.

dariusf commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -93,4 +99,30 @@ public void workAllocationBoardAutoCreationTest() {
+
+        PlatformEx.waitOnFxThread();
+        assertNodeExists(hasText(SAMPLE_BOARD + " has been created and loaded.\n\n" +
+                        "It is saved under the name \"" + SAMPLE_BOARD + "\"."));
```
Same here.

Also, this can be replaced with `waitUntilNodeAppears`. `Platform.runLater` shouldn't be used like that.

codemanmeet commented:
@dariusf Made the changes!

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -23,6 +25,27 @@
+    private static final String SECOND_SAMPLE_REPO_NAME = "HubTurbo/SampleRepo2";
+    public static final String SAMPLE_BOARD_DIALOG =
+            String.format("%s has been created and loaded.%n%nIt is saved under the name \"%s\".",
+                    SAMPLE_BOARD, SAMPLE_BOARD);
```
This message seems a bit redundant. Don't think the second sentence is needed.

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -23,6 +25,27 @@
+
+    public static final List<String> SAMPLE_PANEL_FILTERS =
+            Collections.unmodifiableList(Arrays.asList(
+            String.format("(repo:%s | repo:%s) is:issue sort:updated,comments",
```
We can use the shorthand syntax here.

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -23,6 +25,27 @@
+            String.format("(repo:%s | repo:%s) is:issue sort:updated,comments",
+                    FIRST_SAMPLE_REPO_NAME, SECOND_SAMPLE_REPO_NAME),
+            String.format("repo:%s count:3 is:issue is:open label:\"urgent\"", FIRST_SAMPLE_REPO_NAME),
+            String.format("repo:%s is:issue is:open (as:dariusf || as:codemanmeet)", FIRST_SAMPLE_REPO_NAME)
```
I feel like we shouldn't abbreviate `assignee`. Might not be so intuitive to first-time readers.

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -101,6 +127,26 @@ private void createWorkAllocationBoard() {
+
+        panelData.add(new PanelInfo(SAMPLE_PANEL_NAMES.get(0), SAMPLE_PANEL_FILTERS.get(0)));
+        panelData.add(new PanelInfo(SAMPLE_PANEL_NAMES.get(1), SAMPLE_PANEL_FILTERS.get(1)));
+        panelData.add(new PanelInfo(SAMPLE_PANEL_NAMES.get(2), SAMPLE_PANEL_FILTERS.get(2)));
```
Use a loop for these? For tests too. That was the reason for changing to an array.

damithc commented:
I propose these (I use similar filters in my default board):
11. `Recent unread updates from my repos` -> `(repo:HubTurbo/SampleRepo1 OR repo:HubTurbo/SampleRepo2) is:unread updated<24 sort:~updated`
12. `Progress of the current milestone` -> `repo:Turbo/SampleRepo1 m:curr sort:status`
13. `Issues awaiting prioritization` -> 'repo:Turbo/SampleRepo1 is:open is:issue ~label:priority`

We can add more later. There's no need to limit to 3.

codemanmeet commented:
Sure, will add these three filters to the corresponding issue #1211 for future addition!

damithc commented:
Urm, I meant the 11-13 should be the ones to go for the initial release. More can be added later.

codemanmeet commented:
Oh ok! sure will add these three in this release!

codemanmeet commented:
@dariusf commits squashed and CI passes. will go ahead and merge this now :+1:
